
## Final Project

## Task 1 - Data Loading & Preparation

In [3]:
# we are importing pandas
import pandas as pd

In [4]:
# Loading csv files data to pandas dataframes 
data1=pd.read_csv("data_1_env.csv", sep=",")
data2=pd.read_csv("data_2_env.csv", sep=",")
data3=pd.read_csv("data_3_crop.csv", sep=",")

In [5]:
data1.head()

Year      Location  Avg Min Temp *F  Yr Rain inches Heat Wave Dry Spell  \
0  1980  Andersonport            68.18         27.1280         N         N   
1  1980  Williamhaven            73.40         24.2866         N         N   
2  1980    Josephfurt            70.52         45.8524         N         N   
3  1981  Andersonport            68.00         32.0374         N         n   
4  1981  Williamhaven            74.12         22.2898         N         N   

  Cold Wave Wet Spell  
0         Y         N  
1         Y         N  
2         N         N  
3         Y         N  
4         N         N

In [6]:
data2.head()

Year      Location  Avg Min Temp *C  Yr Rain mm Heat Wave Dry Spell  \
0  1993  Williamhaven             23.6      715.00         N         0   
1  1993    Josephfurt             21.6     1513.69         N         N   
2  1993  Andersonport             21.0     1104.47         N         N   
3  1994  Williamhaven             23.8      787.49         N         N   
4  1994    Josephfurt             21.7     1553.87         N         N   

  Cold Wave Wet Spell  
0         N         N  
1         N         N  
2         N         N  
3         N         N  
4         N         N

In [7]:
data3.head()

Year      Location Irrigation Crop Type  Crop Damage    Yield  \
0  1980.0  Andersonport          Y     Maize          1.0  1311.32   
1  1980.0  Williamhaven          Y     Wheat          1.0  1025.88   
2  1980.0    Josephfurt          N     Maize          0.0  1194.53   
3  1981.0  Andersonport          Y      Oats          1.0  1608.54   
4  1981.0  Williamhaven          Y     Wheat          0.0  1035.73   

           Observer  
0  Marissa Peterson  
1   Frank Carpenter  
2    Kerry Martinez  
3  Marissa Peterson  
4     Eduardo Brady

In [8]:
# Converting Avg Min Temp *F column to Avg Min Temp *C and dropping Avg Min Temp *F column
data1['Avg Min Temp *C'] = data1['Avg Min Temp *F'].apply(lambda x: (x - 32) / 1.8)
data1.drop('Avg Min Temp *F', axis=1, inplace=True)

In [9]:
# Convert Yr Rain inches column to Yr Rain mm and dropping Yr Rain inches columnc
data1['Yr Rain mm'] = (data1['Yr Rain inches'] * 25.4).round(2)
data1.drop('Yr Rain inches', axis=1, inplace=True)

In [10]:
# Defining the desired column order
column_order = ['Year', 'Location', 'Avg Min Temp *C', 'Yr Rain mm','Heat Wave', 'Dry Spell', 'Cold Wave', 'Wet Spell']

#  DataFrame with the new column order
data1 = data1.reindex(columns=column_order)

In [11]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39 entries, 0 to 38
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             39 non-null     int64  
 1   Location         39 non-null     object 
 2   Avg Min Temp *C  39 non-null     float64
 3   Yr Rain mm       39 non-null     float64
 4   Heat Wave        39 non-null     object 
 5   Dry Spell        39 non-null     object 
 6   Cold Wave        39 non-null     object 
 7   Wet Spell        39 non-null     object 
dtypes: float64(2), int64(1), object(5)
memory usage: 2.6+ KB


In [12]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 81 entries, 0 to 80
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             81 non-null     int64  
 1   Location         81 non-null     object 
 2   Avg Min Temp *C  81 non-null     float64
 3   Yr Rain mm       81 non-null     float64
 4   Heat Wave        81 non-null     object 
 5   Dry Spell        81 non-null     object 
 6   Cold Wave        81 non-null     object 
 7   Wet Spell        81 non-null     object 
dtypes: float64(2), int64(1), object(5)
memory usage: 5.2+ KB


### data1 set is having 39 rows and 8 columns and data2 set is having 81 rows and 8 columns, concatening both sets as data 

In [13]:
# concatenating data1 and data2 
data=pd.concat([data1, data2])

In [14]:
data.head()

Year      Location  Avg Min Temp *C  Yr Rain mm Heat Wave Dry Spell  \
0  1980  Andersonport             20.1      689.05         N         N   
1  1980  Williamhaven             23.0      616.88         N         N   
2  1980    Josephfurt             21.4     1164.65         N         N   
3  1981  Andersonport             20.0      813.75         N         n   
4  1981  Williamhaven             23.4      566.16         N         N   

  Cold Wave Wet Spell  
0         Y         N  
1         Y         N  
2         N         N  
3         Y         N  
4         N         N

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 120 entries, 0 to 80
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             120 non-null    int64  
 1   Location         120 non-null    object 
 2   Avg Min Temp *C  120 non-null    float64
 3   Yr Rain mm       120 non-null    float64
 4   Heat Wave        120 non-null    object 
 5   Dry Spell        120 non-null    object 
 6   Cold Wave        120 non-null    object 
 7   Wet Spell        120 non-null    object 
dtypes: float64(2), int64(1), object(5)
memory usage: 8.4+ KB


### The resultant dataset is having 120 rows and 8 columns

In [16]:
# verifying resultant dataset data
data["Location"].value_counts()

Location
Andersonport    40
Williamhaven    40
Josephfurt      40
Name: count, dtype: int64

In [17]:
data["Heat Wave"].value_counts()

Heat Wave
N      80
Y      37
0       1
Yes     1
1       1
Name: count, dtype: int64

In [18]:
#converting '0' to 'Y', Yes to 'Y' and 1 to 'Y' using replace function

data['Heat Wave'] = data['Heat Wave'].replace('Yes', 'Y') 
data['Heat Wave'] = data['Heat Wave'].replace('0', 'N')
data['Heat Wave'] = data['Heat Wave'].replace('1', 'Y') 

In [19]:
# Heat wave have only 'Y' and 'N' after modifying the data
data["Heat Wave"].value_counts()

Heat Wave
N    81
Y    39
Name: count, dtype: int64

In [20]:
data["Dry Spell"].value_counts()

Dry Spell
N     105
0       5
Y       5
No      2
n       1
no      1
y       1
Name: count, dtype: int64

In [21]:
#converting '0' to 'N', No to 'N', y to 'Y', no to 'N' and n to 'N' using replace function

data['Dry Spell'] = data['Dry Spell'].replace('No', 'N') 
data['Dry Spell'] = data['Dry Spell'].replace('no', 'N') 
data['Dry Spell'] = data['Dry Spell'].replace('n', 'N') 
data['Dry Spell'] = data['Dry Spell'].replace('0', 'N') 
data['Dry Spell'] = data['Dry Spell'].replace('y', 'Y') 

In [22]:
# Dry spell have only 'Y' and 'N' after modifying the data
data["Dry Spell"].value_counts()

Dry Spell
N    114
Y      6
Name: count, dtype: int64

In [23]:
data["Cold Wave"].value_counts()

Cold Wave
N    97
Y    20
0     2
n     1
Name: count, dtype: int64

In [24]:
#converting '0' to 'N' and n to 'N' using replace function

data['Cold Wave'] = data['Cold Wave'].replace('n', 'N')
data['Cold Wave'] = data['Cold Wave'].replace('0', 'N')

In [25]:
# Cold Wave have only 'Y' and 'N' after modifying the data
data["Cold Wave"].value_counts()

Cold Wave
N    100
Y     20
Name: count, dtype: int64

In [26]:
data["Wet Spell"].value_counts()

Wet Spell
Y    60
N    58
n     2
Name: count, dtype: int64

In [27]:
#converting n to 'N' using replace function

data['Wet Spell'] = data['Wet Spell'].replace('n', 'N')

In [28]:
# Wet Spell have only 'Y' and 'N' after modifying the data
data["Wet Spell"].value_counts()

Wet Spell
N    60
Y    60
Name: count, dtype: int64

### Now verifying data3 column for any missing values, duplicates and out of range values

In [29]:
data3.head()

Year      Location Irrigation Crop Type  Crop Damage    Yield  \
0  1980.0  Andersonport          Y     Maize          1.0  1311.32   
1  1980.0  Williamhaven          Y     Wheat          1.0  1025.88   
2  1980.0    Josephfurt          N     Maize          0.0  1194.53   
3  1981.0  Andersonport          Y      Oats          1.0  1608.54   
4  1981.0  Williamhaven          Y     Wheat          0.0  1035.73   

           Observer  
0  Marissa Peterson  
1   Frank Carpenter  
2    Kerry Martinez  
3  Marissa Peterson  
4     Eduardo Brady

In [30]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125 entries, 0 to 124
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         124 non-null    float64
 1   Location     124 non-null    object 
 2   Irrigation   124 non-null    object 
 3   Crop Type    124 non-null    object 
 4   Crop Damage  124 non-null    float64
 5   Yield        124 non-null    float64
 6   Observer     124 non-null    object 
dtypes: float64(3), object(4)
memory usage: 7.0+ KB


In [31]:
# Finding missing values in Year column and filling it with 0
data3['Year'] = data3['Year'].fillna(0)

In [32]:
# converting year column in data3 to integer from float datatype

data3['Year'] = data3['Year'].astype(int)

In [33]:
data3["Year"].value_counts()

Year
1991    4
1996    4
1986    4
1988    4
1980    3
2004    3
2005    3
2006    3
2007    3
2008    3
2009    3
2010    3
2011    3
2002    3
2012    3
2013    3
2014    3
2015    3
2016    3
2017    3
2018    3
2003    3
1999    3
2001    3
2000    3
1982    3
1983    3
1984    3
1985    3
1987    3
1989    3
1990    3
1992    3
1993    3
1994    3
1995    3
1997    3
1998    3
1981    3
2019    3
0       1
Name: count, dtype: int64

In [34]:
# Removing the row that is having missing value in year column
data3 = data3[data3['Year'] != 0]

In [35]:
# Removing dupicate values by taking unique combinations of year and location and there are 4 duplciates
data3 = data3.drop_duplicates(subset=['Year', 'Location'])

In [36]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 120 entries, 0 to 124
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         120 non-null    int32  
 1   Location     120 non-null    object 
 2   Irrigation   120 non-null    object 
 3   Crop Type    120 non-null    object 
 4   Crop Damage  119 non-null    float64
 5   Yield        120 non-null    float64
 6   Observer     120 non-null    object 
dtypes: float64(2), int32(1), object(4)
memory usage: 7.0+ KB


In [37]:
data["Location"].value_counts()

Location
Andersonport    40
Williamhaven    40
Josephfurt      40
Name: count, dtype: int64

In [38]:
data3["Irrigation"].value_counts()

Irrigation
Y      78
N      40
Yes     1
1       1
Name: count, dtype: int64

In [39]:
#converting '1' to 'Y' and yes to 'Y' using replace function

data3['Irrigation'] = data3['Irrigation'].replace('Yes', 'Y')
data3['Irrigation'] = data3['Irrigation'].replace('1', 'Y')

In [40]:
# Irrigation column have only 'Y' and 'N' after modifying the data

data3["Irrigation"].value_counts()

Irrigation
Y    80
N    40
Name: count, dtype: int64

In [41]:
data3["Crop Type"].value_counts()

Crop Type
Wheat    47
Maize    37
Oats     36
Name: count, dtype: int64

In [42]:
# There is one missing value in crop damage column and filling it with 0
data3['Crop Damage'] = data3['Crop Damage'].fillna(0)

In [43]:
data3["Crop Damage"].value_counts()

Crop Damage
0.0    87
1.0    33
Name: count, dtype: int64

In [44]:
#converting '0.0' to 'N' and '1.0' to 'Y' using replace function

data3['Crop Damage'] = data3['Crop Damage'].replace(0.0, 'N')
data3['Crop Damage'] = data3['Crop Damage'].replace(1.0, 'Y')

In [45]:
# Irrigation column have only 'Y' and 'N' after modifying the data

data3["Crop Damage"].value_counts()

Crop Damage
N    87
Y    33
Name: count, dtype: int64

In [46]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 120 entries, 0 to 124
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         120 non-null    int32  
 1   Location     120 non-null    object 
 2   Irrigation   120 non-null    object 
 3   Crop Type    120 non-null    object 
 4   Crop Damage  120 non-null    object 
 5   Yield        120 non-null    float64
 6   Observer     120 non-null    object 
dtypes: float64(1), int32(1), object(5)
memory usage: 7.0+ KB


### Now there are no missing values, duplicate values and out of range values in Data3, it has 120 rows and 7 columns

### Merging data3 and data, naming it as clean_data

In [47]:
clean_data = pd.merge(data, data3, on=['Year', 'Location'])

### clean_data file has 120 rows and 13 columns of data with no null values and out of range values

In [49]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             120 non-null    int64  
 1   Location         120 non-null    object 
 2   Avg Min Temp *C  120 non-null    float64
 3   Yr Rain mm       120 non-null    float64
 4   Heat Wave        120 non-null    object 
 5   Dry Spell        120 non-null    object 
 6   Cold Wave        120 non-null    object 
 7   Wet Spell        120 non-null    object 
 8   Irrigation       120 non-null    object 
 9   Crop Type        120 non-null    object 
 10  Crop Damage      120 non-null    object 
 11  Yield            120 non-null    float64
 12  Observer         120 non-null    object 
dtypes: float64(3), int64(1), object(9)
memory usage: 12.3+ KB


In [48]:
clean_data.isna().sum()

Year               0
Location           0
Avg Min Temp *C    0
Yr Rain mm         0
Heat Wave          0
Dry Spell          0
Cold Wave          0
Wet Spell          0
Irrigation         0
Crop Type          0
Crop Damage        0
Yield              0
Observer           0
dtype: int64

In [50]:
# Saving the dataset to the clean_data
clean_data.to_csv('clean_data.csv', index=False)

In [51]:
!jupyter nbconvert --to html task_1.ipynb

[NbConvertApp] Converting notebook task_1.ipynb to html
[NbConvertApp] Writing 651723 bytes to task_1.html
